File: hackerrank/problem solving/algorithms/2 - implementation/32 - sherlock and squares/sherlock_and_squares.rb
# https://www.hackerrank.com/challenges/sherlock-and-squares/problem?isFullScreen=true

def main
  n = gets.to_i
  n.times do
    a, b = read_numbers
    puts squares(a, b)
  end
end

def read_numbers
  gets&.split&.map!(&:to_i)
end

# T: O(1)
# S: O(1) extra space
def squares(start_num, end_num)
  max_square = Math.sqrt(end_num).floor.to_i
  min_square = Math.sqrt(start_num).ceil.to_i
  max_square - min_square + 1
end

main if __FILE__ == $0

File: hackerrank/problem solving/algorithms/2 - implementation/32 - sherlock and squares/sherlock_and_squares.cpp
// https://www.hackerrank.com/challenges/sherlock-and-squares/problem?isFullScreen=true

#include <cmath>
#include <iostream>

using namespace std;

int squares(int start_num, int end_num);

int main()
{
    int n;
    cin >> n;
    for (int i{}; i < n; ++i) {
        int a, b;
        cin >> a >> b;
        cout << squares(a, b) << '\n';
    }
}

int squares(int start_num, int end_num)
{
    int max_square = floor(sqrt(end_num));
    int min_square = ceil(sqrt(start_num));
    return max_square - min_square + 1;
}
